pkg/stanza/fileconsumer/reader.go

```diff
@@ -85,23 +94,93 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 		}
 
 		token, err := r.encoding.Decode(scanner.Bytes())
-		if err != nil {
+		switch {
+		case err != nil:
 			r.Errorw("decode: %w", zap.Error(err))
-		} else {
+		case r.headerSettings != nil && !r.HeaderFinalized:
+			if !r.consumeHeaderLine(ctx, r.fileAttributes, token) {
+				// recreate the scanner with the log-line's split func.
+				// We do not use the updated offset from the scanner,
+				// as the log line we just read could be multiline, and would be
+				// split differently with the new splitter.
+				if _, err := r.file.Seek(r.Offset, 0); err != nil {
+					r.Errorw("Failed to seek post-header", zap.Error(err))
+					return
+				}
+
+				scanner = NewPositionalScanner(r, r.maxLogSize, r.Offset, r.curSplitFunc())
+			}
```

Comment:
Isn't it possible to initialize the reader with the appropriate function, and then in the case where we consumer a header, just update the function once when the header is finalized?

Reply:
I'll look into it, but my reasoning for doing it this way:

We need to remake the scanner (which is local to ReadToEnd) when the header finalizes, because the split function may differ, so we need to pass state upwards out of the emit function in order to do that.

I can try to track if the state transitions from headerFinalized = false -> headerFinalized = true and remake the scanner then, but if our goal is to make this clearer, I'm not particularly convinced that refactoring like this will help.

Reply:
Implemented as stated above ^ let me know if you have a better way to remake that scanner, I'd definitely like to factor that logic out of the read loop!